lib/starttls.js

```diff
@@ -33,111 +15,46 @@ function starttlsClient(socket, options, callback) {
 }
 
 function starttls(socket, options, callback, isServer) {
-  var sslcontext;
-
   var opts = {};
+
   Object.keys(options).forEach(function(key) {
     opts[key] = options[key];
   });
   if (!opts.ciphers) {
     opts.ciphers = RECOMMENDED_CIPHERS;
   }
+  opts.isServer = isServer;
+  opts.secureContext = tls.createSecureContext(opts);
 
-  socket.removeAllListeners('data');
-  if (tls.createSecureContext) {
-    sslcontext = tls.createSecureContext(opts);
-  } else {
-    sslcontext = crypto.createCredentials(opts);
-  }
-  var pair = tls.createSecurePair(sslcontext, isServer);
-  var cleartext = pipe(pair, socket);
-
+  var secureSocket = new tls.TLSSocket(socket, opts);
   var erroredOut = false;
-  pair.on('secure', function() {
+
+  // NodeJS documentation bug: secure vs secureConnect
+  // https://github.com/nodejs/node/issues/10555
+  secureSocket.on('secure', function() {
     if (erroredOut) {
-      pair.end();
+      secureSocket.end();
       return;
     }
 
-    var verifyError = (pair._ssl || pair.ssl).verifyError();
-
-    if (verifyError) {
-      cleartext.authorized = false;
-      cleartext.authorizationError = verifyError;
+    var authError = secureSocket.ssl.verifyError();
+    if (authError) {
+        secureSocket.authorized = false;
+        secureSocket.authorizationError = authError;
     } else {
-      cleartext.authorized = true;
+        secureSocket.authorized = true;
```

Comment:
@mk-pmb: thanks for your reply!  It is debatable if we should provide our own style of authenticating clients which is different from existing standard TLS options. With standard TLS module there are two options which influence what we discuss here: `requestCert` and `rejectUnauthorized`. In ftpd's documentation we claim that `tlsOptions` are standard arguments of tls.createServer. It can get very confusing if we introduce our own new option `requireTlsCert` which overlaps in function with existing TLS options (in particular when options are not consistent).

I would rather suggest to leverage existing TLS options and only make checkServerIdentity option a bit different because that is the place where special needs of FTP protocol need to be taken into account. I agree with all info which needs to be passed to that function as you have outlined that above (validation event arguments). I would just suggest to keep everything the same as in case of standard checkServerIdentity function and only pass additional argument(s) to it with FTP session information etc. (so a validation callback instead of an "event"). `allowUnauthorizedTls` option should be removed entirely (in meaning it is the same as TLS `rejectUnauthorized` option so no need for it).

However I also have to say that I don't have time to work on this feature (at least now) and it's getting a bit far from the original bug which I'm trying to fix here (which is accidental file truncation upon upload). But I would like to prepare the ground for future enhancements in this area so that we get one step closer to ideal solution (for now that would mean calling checkServerIdentity function passing `socket.remoteAddress` instead of hostname to it). How does that sound to you?